File: frontend/src/protocol/mangekyou.ts
import type { ChatCompletionRequest } from './openai'

export interface MangekyouRequestBody {
    api_url: string
    api_key: string
    openai_request: ChatCompletionRequest
}

export interface MangekyouRequest extends MangekyouRequestBody {
    id: number
}

export interface MangekyouResponseBase {
    id: number
}

export interface MangekyouSuccessResponse extends MangekyouResponseBase {
    content: string
    prompt_tokens: number
    completion_tokens: number
}

export interface MangekyouStreamChunk extends MangekyouResponseBase {
    delta: string
}

export interface MangekyouStreamEnd extends MangekyouResponseBase {
    done: boolean
    prompt_tokens?: number
    completion_tokens?: number
}

export interface MangekyouErrorResponse extends MangekyouResponseBase {
    error: string
}

export type MangekyouResponse = MangekyouSuccessResponse | MangekyouErrorResponse | MangekyouStreamChunk | MangekyouStreamEnd

File: backstage/src/protocol/mangekyou.rs
use serde::{Serialize, Deserialize};

use crate::protocol::openai;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MangekyouRequest {
    pub id: u32,

    pub api_url: String,
    pub api_key: String,

    pub openai_request: openai::ChatCompletionRequest
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MangekyouSuccessResponse {
    pub id: u32,

    pub content: String,
    pub prompt_tokens: u32,
    pub completion_tokens: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MangekyouStreamChunk {
    pub id: u32,
    pub delta: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MangekyouStreamEnd {
    pub id: u32,
    pub done: bool,
    pub prompt_tokens: Option<u32>,
    pub completion_tokens: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MangekyouErrorResponse {
    pub id: u32,
    pub error: String
}
